src/merkle/store.rs

```diff
@@ -0,0 +1,375 @@
+//! An in-memory data store for Merkle-lized data
+//!
+//! This is a in memory data store for Merkle trees, this store allows all the nodes of a tree
+//! (leaves or internal) to live as long as necessary and without duplication, this allows the
+//! implementation of efficient persistent data structures
+use super::{
+    BTreeMap, Direction, MerkleError, MerklePath, MerkleTree, NodeIndex, Rpo256, RpoDigest,
+    SimpleSmt, Vec, Word,
+};
+
+// FIX: This is a temporary work around for the lack of Ord in BaseElement
+#[repr(transparent)]
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+struct WordWithOrd {
+    word: Word,
+}
+
+impl PartialOrd for WordWithOrd {
+    fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
+        let self_ints = self.word.iter().map(|e| e.inner());
+        let other_ints = other.word.iter().map(|e| e.inner());
+
+        self_ints.partial_cmp(other_ints)
+    }
+}
+
+impl Ord for WordWithOrd {
+    fn cmp(&self, other: &Self) -> core::cmp::Ordering {
+        let self_ints = self.word.iter().map(|e| e.inner());
+        let other_ints = other.word.iter().map(|e| e.inner());
+
+        self_ints.cmp(other_ints)
+    }
+}
+
+impl From<Word> for WordWithOrd {
+    fn from(word: Word) -> Self {
+        WordWithOrd { word }
+    }
+}
+
+impl From<WordWithOrd> for Word {
+    fn from(word: WordWithOrd) -> Self {
+        word.word
+    }
+}
+
+impl From<RpoDigest> for WordWithOrd {
+    fn from(digest: RpoDigest) -> Self {
+        WordWithOrd {
+            word: digest.into(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Node {
+    left: WordWithOrd,
+    right: WordWithOrd,
+}
+
+pub struct MerkleStore {
+    nodes: BTreeMap<WordWithOrd, Node>,
+}
+
+impl Default for MerkleStore {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl MerkleStore {
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+    pub fn new() -> MerkleStore {
+        let mut nodes = BTreeMap::new();
+        let smt = SimpleSmt::new(SimpleSmt::MAX_DEPTH)
+            .expect("MAX_DEPTH must be a valid depth for a SMT");
+
+        // pre-populate the store with the empty hashes
+        for empty in smt.store.empty_hashes {
+            let hash = Rpo256::merge(&[empty, empty]);
+            nodes.insert(
+                hash.into(),
+                Node {
+                    left: empty.into(),
+                    right: empty.into(),
+                },
+            );
+        }
```

Comment:
We probably just want the [EmptyRoots](https://github.com/0xPolygonMiden/crypto/blob/2871e4eb27be074f2cc80881342d0bb6d15929f8/src/merkle/empty_roots.rs#L17-L24) here

Reply:
Yeah, I wondered why the simple smt has the logic to overwrite the empty values and why it was using the constant values, any ideas?

Reply:
The principle behind the SimpleSMT is to allow us to use compound trees to compose more complex data structures (i.e. Tiered SMT).

Example: if we use a SimpleSMT to cover the second tier, then its root will start at depth `16` (so the empty roots set should be relative to `16`)

Reply:
Yes, here we should just insert empty roots rather than re-compute them.

Reply:
done